ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-commands.sh

```diff
@@ -4,13 +4,30 @@ set -o nounset
 set -o errexit
 set -o pipefail
 
-
-
 # We have to truncate cluster name to 14 chars, because there is a limitation in the install-config
 # Now it looks like "ci-op-rl6z646h-65230".
 # We will remove "ci-op-" prefix from there to keep just last 14 characters. and it cannot start with a "-"
 UNSAFE_CLUSTER_NAME=${NAMESPACE}-${JOB_NAME_HASH}
 SAFE_CLUSTER_NAME=${UNSAFE_CLUSTER_NAME#"ci-op-"}
-echo "${SAFE_CLUSTER_NAME}" > ${SHARED_DIR}/CLUSTER_NAME
 
+case "$CLUSTER_TYPE" in
+  openstack-vexxhost)
+    OS_CLIENT_CONFIG_FILE="$OS_CLIENT_CONFIG_FILE_VEXXHOST"
+    OPENSTACK_EXTERNAL_NETWORK='public'
+    OPENSTACK_INSTANCE_FLAVOR='v1-standard-4'
+    ;;
+  openstack)
+    OS_CLIENT_CONFIG_FILE="$OS_CLIENT_CONFIG_FILE_MOC"
+    OPENSTACK_EXTERNAL_NETWORK='external'
+    OPENSTACK_INSTANCE_FLAVOR='m1.s2.xlarge'
+    ;;
+  *)
+    echo >&2 "Unsupported cluster type '${CLUSTER_TYPE}'"
+    exit 1
+    ;;
+esac
 
+cp   "$OS_CLIENT_CONFIG_FILE"        "${SHARED_DIR}/clouds.yaml"
+echo "$OPENSTACK_EXTERNAL_NETWORK" > "${SHARED_DIR}/OPENSTACK_EXTERNAL_NETWORK"
+echo "$OPENSTACK_INSTANCE_FLAVOR"  > "${SHARED_DIR}/OPENSTACK_INSTANCE_FLAVOR"
+echo "$SAFE_CLUSTER_NAME"          > "${SHARED_DIR}/CLUSTER_NAME"
```

Comment:
> I think what you should be moving towards is removing the use of "CLUSTER_TYPE"

The step-registry configuration is designed around the use of `CLUSTER_TYPE`: for example, leases depend on that variable. In the generic Install step, `CLUSTER_TYPE` is used to expose the cloud credentials to the installer. I don't think it's a good idea to depart from that convention.

Reply:
> for example, what if I want to run on vexxhost but I want to use v2-standard-3" flavor?

This is a very good example, because it lets me highlight a principle I am following in this patch:

**Flavor names are a cloud-specific implementation detail.**

The reason why flavor names only appear once after this patch, and in a `.sh` file, is because they are arbitrary and each cloud has its naming convention, which has nothing to do with the test case definition.

With this patch, cloud-specific configuration (e.g. flavor names and external network name) is not part of the API of the test any more.

I believe that the test you propose should not be defined as "I want to use Vexxhost and I want to use 'v2-standard-3'". Rather, it would be: "I want to test on machines with just 8GB RAM".

When we want to implement that test, we could make `ipi-conf-openstack` accept a new environment variable, containing the requested RAM. The script will then map the requested RAM value to the appropriate flavor name for each cloud, if it exists. This way, we keep a clear separation between business objects (e.g. RAM quantity) and implementation detail (e.g. flavor names). You won't have to change the test definition when you move across cloud providers.

Reply:
@adduarte you bring up a good point. We want to retain the flexibility of setting these variables individually per workflow but at the same time, make it easy to switch workflows from one cloud platform to another, by hiding the complexity behind `CLUSTER_TYPE`.

How about we set default values to the above variables, depending on `CLUSTER_TYPE`?

```
case "$CLUSTER_TYPE" in
  openstack-vexxhost)
    OPENSTACK_EXTERNAL_NETWORK==${OPENSTACK_EXTERNAL_NETWORK:-'public'}
    OPENSTACK_INSTANCE_FLAVOR==${OPENSTACK_INSTANCE_FLAVOR:-'v1-standard-4'}
    ;;
  openstack)
    OPENSTACK_EXTERNAL_NETWORK==${OPENSTACK_EXTERNAL_NETWORK:-'external'}
    OPENSTACK_INSTANCE_FLAVOR==${OPENSTACK_INSTANCE_FLAVOR:-'m1.s2.xlarge'}
    ;;
  *)
    echo >&2 "Unsupported cluster type '${CLUSTER_TYPE}'"
    exit 1
    ;;
esac
```

I'm specifically leaving `OS_CLIENT_CONFIG_FILE` and `OS_CLOUD` out, because I think we should standardize their values.

Reply:
I think that at the workflow level, we should set "a specific amount of RAM", and not "a specific flavor name".

This can be achieved in multiple ways.

## Solution 1
We use `ipi-conf-openstack` as a mapping that finds the right flavor name based on variables

## Solution 2
We create `ipi-conf-openstack-low-memory`, `ipi-conf-openstack-big-disk` etc, and include them in pre-chains to be explicitly called in the workflow definition.

Example (in `ci-operator/config/openshift/installer/openshift-installer-master.yaml`):
```yaml
- artifact_dir: /tmp/artifacts
  as: e2e-openstack-ipi-low-memory
  steps:
    cluster_profile: openstack
    workflow: openshift-e2e-openstack-ipi
    pre:
    - chain: ipi-pre-openstack-pre-lowram
```

---

The point is: there are multiple effective ways of solving problems, while maintaining a separation between business logic ("how much RAM I want to test on") and implementation detail ("the flavor name"). Once we've proven that the concept holds, we don't need to implement every future potential need in this PR.

Reply:
RAM = Flavors. They are the same thing. in OpenStack a user defines how much ram they want to use by specifying the flavor. 
That's why so often you see the flavor names listing the RAM in the name. The user defines the flavors. 
What we should really do is standarize on the names we used when we define flavors in a cloud provider. 
But its irrelevant. 

**Solution 3: **
you define it in the workflow where it is easy visible to whoever is looking at the definition. 
like you say : 

"I want a specific amount of RAM in this CLUSTER_TYPE" is the same thing as "flavor: flavor_with_ram_I_want" 
Those two are really just strings, but with the first one you have to do some sort of lookup and translating it to "flavor: flavor_with_ram_i_want" 

I like what @mandre is saying. define defaults but keep it open so I can override them via the workflow, and *only* the workflow file. 

That should be our goal to. we should be able to define a workflow (new test) by simply adding one single file, the workflow file.

Reply:
> RAM = Flavors. They are the same thing

I disagree. RAM has business value. "Flavor name" is an implementation detail we abstract away in this patch.

> standarize on the names we used when we define flavors in a cloud provider

This has been discussed in private (sorry) with @mandre. We settled on: "flavors are arbitrary names and we can't assume we can normalise them across clouds".

> I want a specific amount of RAM in this CLUSTER_TYPE

The Boskos lease depends on the CLUSTER_TYPE. This means that CLUSTER_TYPE **must** represent a cloud; not a configuration.

I agree that we should use environment variables to define the RAM; we will code that in a follow-up of this patch, once we have that need.